c2rust-transpile/src/translator/mod.rs

```diff
@@ -3110,21 +3124,32 @@ impl<'c> Translation<'c> {
         &self,
         ctx: ExprContext,
         type_id: CTypeId,
+        override_ty: Option<CQualTypeId>,
     ) -> TranslationResult<WithStmts<Box<Expr>>> {
         if let CTypeKind::VariableArray(elts, len) = self.ast_context.resolve_type(type_id).kind {
             let len = len.expect("Sizeof a VLA type with count expression omitted");
 
-            let elts = self.compute_size_of_type(ctx, elts)?;
+            let elts = self.compute_size_of_type(ctx, elts, override_ty)?;
             return elts.and_then(|lhs| {
-                let len = self.convert_expr(ctx.used().not_static(), len)?;
+                let len = self.convert_expr(ctx.used().not_static(), len, override_ty)?;
                 Ok(len.map(|len| {
                     let rhs = cast_int(len, "usize", true);
                     mk().binary_expr(BinOp::Mul(Default::default()), lhs, rhs)
                 }))
             });
         }
         let ty = self.convert_type(type_id)?;
-        self.mk_size_of_ty_expr(ty)
+        let mut result = self.mk_size_of_ty_expr(ty);
+        // cast to expected ty if one is known
+        if let Some(expected_ty) = override_ty {
+            trace!(
+                "Converting result of sizeof to {:?}",
+                self.ast_context.resolve_type(expected_ty.ctype)
+            );
+            let result_ty = self.convert_type(expected_ty.ctype)?;
+            result = result.map(|x| x.map(|x| mk().cast_expr(x, result_ty)));
```

Comment:
Doesn't this introduce more casts?

Could the value already be the expected type?

Reply:
`std::mem::size_of` always produces a `usize`, so the only case where this cast is not necessary is when the surrounding expression already wants a `usize` or something type-aliased to it.

We never directly generate `usize` in transpiled code, only `size_t` which is a type alias to it. We could remove the cast in that case, but leaving it in is more flexible in case the user wants to alter the definition of the `size_t` type alias later on to something that would require explicit casts from `usize`.